### PR TITLE
Add POCL support and CI + major package rework

### PR DESCRIPTION
**Pull request by maleadt**
This is a PR to add POCL support to OpenCL.jl, since I may want to use it to implement the CPU back-end of KernelAbstractions.jl. Since there are ~very few~ no real public packages depending on OpenCL.jl, I'm additionally doing a rework/modernization of the package in the hopes to revitalize it a bit, and bring it closer to the other GPU back-ends we have in Julia. Needless to say, this is going to be breaking.

Breaking changes:

- `Buffer` ctor: switch length and flag arguments around (for ambiguity reasons), and make passing the length mandatory

TODO:

- remove questionable APIs: `create_some_context` etc

### CODE REVIEW

File: .github/workflows/CI.yml [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    name: Julia ${{ matrix.version }} - ${{ matrix.os }} - ${{ matrix.arch }}
+    runs-on: ${{ matrix.os }}
+    strategy:
+      fail-fast: false
+      matrix:
+        version: ['1.10', 'pre']
+        os: ['ubuntu-latest']
+        arch: [x64]
+    steps:
+      - uses: actions/checkout@v2
```
Comment (giordano) [inline .github/workflows/CI.yml:25]:
```suggestion
      - uses: actions/checkout@v4
```

File: .github/workflows/CI.yml [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+          restore-keys: |
+            ${{ runner.os }}-test-${{ env.cache-name }}-
+            ${{ runner.os }}-test-
+            ${{ runner.os }}-
+      - uses: julia-actions/julia-buildpkg@v1
+      - uses: julia-actions/julia-runtest@v1
+      - uses: julia-actions/julia-processcoverage@v1
+      - uses: codecov/codecov-action@v2
+        with:
+          files: lcov.info
```
Comment (giordano) [inline .github/workflows/CI.yml:45]:
```suggestion
      - uses: codecov/codecov-action@v4
        with:
          token: ${{ secrets.CODECOV_TOKEN }}
          files: lcov.info
```
but you need to set up a codecov token for this to work (better if it's an org wide one, so you don't have to create for all individual repos of the org)

Comment (maleadt) [reply .github/workflows/CI.yml:36]: TIL about the org-wide token; hard to find though on the codecov webui.

**Comment by maleadt** [comment]
Yay, works on macOS now too!

**Comment by maleadt** [comment]
And finally CUDA CI. I think I'm going to postpone other refactors to a different PR, so let's merge this.

**Pull request merged**